### DISCUSSION TITLE
Idea: pin-pointing by sound

### DISCUSSION THREAD

**Discussion by eben80** [dl9rdz/rdz_ttgo_sonde#199] [Developers corner]
I believe that ESP32 can create tones by PWM and a piezo buzzer right?

How about making the tracker have a pin-point function like with metal detectors.

ie. On devices with GPS you have the route to sonde parameter (0 means you're heading straight for it)

So how about having the ability to switch on a mode where:
1.  The tone from the buzzer will be at it's highest when you are on 0 degrees (heading straight for sonde) 
2.  The number of tones(like clicks) will increase as the signal strength increases.

So when you're in a corn field for example, you will head in the direction with the highest tone first, and then keep aiming for the highest note and fastest "clicks".

I was lost in a corn field yesterday, which is probably what drove me to thinking of this :D

**Comment by dl9rdz** [comment]
There is one drawback: The TTGO GPS will determine your own heading only if you are moving fast enough. I am not yet convinced this will work good enough in practice.

There has been a different suggestion somewhere else already: Map the RSSI value to a sound. Of course this requires you to have a directional antenna, but then you can use the TTGO for direction finding. Could be great for hunting a RS41SGM :-)

**Reply by eben80** [reply]
Of course that's a good point , you have to be moving. The heading to target number has always been a good guide for me when hunting but the only reason I keep on using an offline map is because its a lot easier than trying to keep it on 0 while walking over rough terrain 😆 
Sound might be better but I get what you're saying.  This is also why I thought having RSSI as a second sound modifier.
The SGM doesn't send any location telemetry? It's the military one right?

**Comment by eben80** [comment]
I want to have a go at this. 
The idea is to use a piezo buzzer connected to a free pin, together with the EasyBuzzer library to change the tone frequency and length, based on COG and RSSI.
`void beep(unsigned int frequency, unsigned int const onDuration, unsigned int const offDuration, byte const beeps, unsigned int const pauseDuration, unsigned int const sequences);`

Where would the best place in the code be to call this from where COG and RSSI values are available and updated regularly? In the display function?

**Comment by eben80** [comment]
OK so I've tried to think of a better way to do this taking into account the esp32 tone library limitations.
Using [Tone32.hpp](https://github.com/makeabilitylab/arduino/blob/master/MakeabilityLab_Arduino_Library/src/Tone32.hpp)
```
#include <Arduino.h>
#include <Tone32.hpp>

const int TONE_OUTPUT_PIN = 13;
const int TONE_PWM_CHANNEL = 0;

Tone32 _tone32(TONE_OUTPUT_PIN, TONE_PWM_CHANNEL);
const int PLAY_NOTE_DURATION_MS = 200;
int TONE_GAP_DURATION = 2000;
int TONE_FREQ = 100;
int RSSI = 15;     // Assumes abs value of around min 15 max 120
int HEADING = 180; // 0 to 359
unsigned long time_1 = 0;
String a;

void toneLocate(int head, int sig)
{
  if (_tone32.isPlaying())
  {
  }
  else
  {
    if (millis() >= time_1 + TONE_GAP_DURATION)
    {
      time_1 += TONE_GAP_DURATION;
      TONE_GAP_DURATION = 200 + sig / 0.1;
      if (head == 0)
      {
        TONE_FREQ = 9200;
      }
      else if (head <= 180)
      {
        TONE_FREQ = (9000 / head);
      }
      else
      {
        TONE_FREQ = (9000 / (360 - head));
      }
      _tone32.playTone(TONE_FREQ, PLAY_NOTE_DURATION_MS);
    }
  }
}

void setup()
{
  Serial.begin(115200);
}

void loop()
{
  while (Serial.available())
  {

    a = Serial.readString(); // read the incoming data as string
    HEADING = a.toInt();
  }

  toneLocate(HEADING, RSSI);

  _tone32.update();
  Serial.print("TONE_GAP_DURATION");
  Serial.println(TONE_GAP_DURATION);
  Serial.println();
  Serial.print("TONE_FREQ");
  Serial.println(TONE_FREQ);
  Serial.println();
  Serial.println();
}
```

As you can see it has a toneLocate function that uses heading and abs rssi.
This expects the ttgo to have gps to provide heading but of course it could also just work on rssi alone.

The closer you are to 0 heading, the higher the tone will be.
The stronger the signal, the more frequent the tones will be.

Now what I don't know is how to implement it in the current loop. @dl9rdz  could you give me a hint?  How do I pass the heading and rssi to the function call in loop?
I also don't know if the _tone32.update being called every time would cause issues because of its polling nature.

Thank you as always.
